py4ai/core/data/layer/v1/pandas/repository.py

```diff
@@ -0,0 +1,222 @@
+                ascending=[
+                    True if sort[1] == SortingDirection.ASC else False
+                    for sort in options.sorting_options
+                ],
+            )
+            if len(options.sorting_options) > 0
+            else rows
+        )
+
+        size = rows.shape[0]
+        elements = [
+            self.serializer.to_model(row)
+            for row_id, row in rows.iloc[
+                options.page_start : (options.page_start + options.page_size)
+                if options.page_size != -1
+                else None
+            ].iterrows()
+        ]
+
+        return Paged(size, elements, options.page_start + options.page_size < size)
+
+    async def create(self, obj: E) -> E:
+        """
+        Insert an object of type Document/pd.DataFrame/pd.Series in a pd.DataFrame.
+
+        :param obj: An instance of :class:`cgnal.data.model.text.Document, pd.DataFrame or pd.Series`
+        :return: self i.e. an instance of ``PandasArchiver`` with updated self.data object
+
+        """
+        return (await self.save([obj]))[0]
+
+    def __create_series_from_entity(self, entity: E) -> pd.Series:
+        serie = self.serializer.to_document(entity)
+        serie.name = self.serializer.to_document_key(self.serializer.get_key(entity))
+        return serie
+
+    async def save(self, entities: Sequence[E]) -> Sequence[E]:
+        """
+        Insert many objects of type Document/pd.DataFrame/pd.Series in a pd.DataFrame.
+
+        :param objs: List of objects to be inserted. The objects can be of the following class instances
+            :class:`cgnal.data.model.text.Document`, `pd.DataFrame` or `pd.Series`
+        :return: self i.e. an instance of ``PandasArchiver`` with updated self.data object
+
+        """
+        new = pd.DataFrame(
+            [self.__create_series_from_entity(entity) for entity in entities]
+        )
+        self.data = pd.concat(
+            [self.data.loc[list(set(self.data.index).difference(new.index))], new]
+        ).sort_index()
+        self.commit()
+        return entities
+
+    async def list(self, options: QueryOptions = QueryOptions()) -> Paged[E]:
+        return await self.retrieve_by_criteria(PandasSearchCriteria.empty(), options)
+
+    async def delete(self, key: KE) -> bool:
+        document_key = self.serializer.to_document_key(key)
+        self.data = self.data.drop(document_key)
+        self.commit()
+        return True
+
+    async def delete_by_criteria(self, criteria: SearchCriteria[Q]) -> bool:
+        rows = self.data[criteria.query(self.data)]
+        self.data = self.data.drop(rows.index)
+        self.commit()
+        return True
+
+
+class CsvRepository(PandasRepository[KE, KD, E]):
+    """Archiver based on persistent layers based on tabular files stored on csv."""
+
+    def __init__(
+        self,
+        filename: PathLike,
+        serializer: DataSerializer[KE, KD, E, pd.Series],
+        sep: str = ";",
```

Comment:
I'll add here also the decimal separator and maybe also other parameters relevant for pd.DataFrame.read_csv, like index_col, usecols, dtype, etc...

Reply:
Uhm, separator is ok, all the other parameters I believe would make the init very messy. I'd leave it simple and neat. If we really have use-cases, these can be either extended or PR against the repo